### ISSUE TITLE
Working memory accuracy

### ISSUE THREAD

**Issue by hunse** [nengo/nengo-loihi#149]
This issue comes out of problems we ran into with the benchmarks. It appears that storing information in working memory (a common feature of SPA models) does not work as well as it should.

What I've found so far is that the behaviour of the working memory is extremely sensitive to some initial conditions (I'm not sure which). Running the network without a seed, I find that occasionally it works well, sometimes it works okay, and sometimes it does not work at all.

I've been testing this using the script below (taken from the ConvolutionCleanup benchmark in `nengo_benchmarks`).
```python
import matplotlib.pyplot as plt
import nengo
import nengo.spa as spa
import nengo_loihi
import numpy as np

D = 2
mem_input_scale = 0.5
mem_tau = 0.1

model = spa.SPA()
with model:
    model.shape = spa.Buffer(D)
    model.color = spa.Buffer(D)
    model.bound = spa.Buffer(D)

    nengo.Connection(model.bound.state.output, model.bound.state.input,
                     synapse=mem_tau)

    cconv = nengo.networks.CircularConvolution(n_neurons=200, dimensions=D)

    nengo.Connection(model.shape.state.output, cconv.A)
    nengo.Connection(model.color.state.output, cconv.B)
    nengo.Connection(cconv.output, model.bound.state.input,
                     transform=mem_input_scale, synapse=mem_tau)

    stim_time = mem_tau / mem_input_scale

    def stim_color(t):
        if 0 < t < stim_time:
            return 'BLUE'
        elif stim_time < t < stim_time*2:
            return 'RED'
        else:
            return '0'

    def stim_shape(t):
        if 0 < t < stim_time:
            return 'CIRCLE'
        elif stim_time < t < stim_time*2:
            return 'SQUARE'
        else:
            return '0'

    model.input = spa.Input(
        shape = stim_shape,
        color = stim_color,
        )

    probe_wm = nengo.Probe(model.bound.state.output, synapse=0.02)

vocab_wm = model.get_output_vocab('bound')

test_time = 2.0
# with nengo.Simulator(model) as sim:
with nengo_loihi.Simulator(model) as sim:
    T = stim_time * 2 + test_time
    sim.run(T)

vocab_wm = vocab_wm
vals_wm = [None] * 2
vals_wm[0] = np.dot(sim.data[probe_wm], vocab_wm.parse('BLUE*CIRCLE').v)
vals_wm[1] = np.dot(sim.data[probe_wm], vocab_wm.parse('RED*SQUARE').v)
vals_wm = np.array(vals_wm)

plt.plot(sim.trange(), vals_wm.T)
plt.legend(['BLUE*CIRCLE', 'RED*SQUARE'], loc='best')
plt.xlabel('time (s)')
plt.ylabel('memory')
plt.show()
```

I'll continue looking into this, but I wanted to start the issue as a place for discussion/collaboration.

**Comment by tcstewar** [comment]
> I can't set all of the connections here to full-weight matrices because many of them go through passthrough nodes. Is there a lower-level way to force this in the builder?

Not really.  I tried hacking this into the builder and it's doable, but ugly (basically search for every place that `solver.weights` gets used).....   But, I think it'd be possible to hack in a change to `splitter.split` that just goes through all the connections (after removing passthroughs) and replaces the solver with a `weights=True` solver.....  https://github.com/nengo/nengo-loihi/blob/master/nengo_loihi/splitter.py#L111

**Comment by arvoelke** [comment]
What branch(es) are you running it on when it works well with a fixed vocab seed? It's not working so well for me with the emulator on `integrator-accuracy`, and with a lot of variability (using `seed=0` for the vocab):

**Trial 1**:
![trial1](https://user-images.githubusercontent.com/5420057/49665975-a3486f00-fa24-11e8-9a88-d98c1590c8b3.png)

**Trial 2**:
![trial2](https://user-images.githubusercontent.com/5420057/49665978-a3e10580-fa24-11e8-92de-00ee160a16bf.png)

**Nengo (Reference)**:
![nengo_wm](https://user-images.githubusercontent.com/5420057/49665927-7ac07500-fa24-11e8-801c-a1dc2c18469f.png)

**Comment by hunse** [comment]
I'm doing everything on #132.

**Comment by arvoelke** [comment]
`add_inputs` is definitely better. Much less variability, and resembles the ideal.

![trial1](https://user-images.githubusercontent.com/5420057/49666644-bceab600-fa26-11e8-8ced-2089bdea2724.png)

![trial2](https://user-images.githubusercontent.com/5420057/49666647-be1be300-fa26-11e8-8e8c-7c0e57aa1315.png)

![trial3](https://user-images.githubusercontent.com/5420057/49666648-bfe5a680-fa26-11e8-8658-2fc12d589228.png)

**Comment by hunse** [comment]
That last plot you posted, that's the kind of behaviour I was getting with the shorter `dt`. I'm not sure how problematic it is, because it could just be because of a bad vocab. Are you able to reproduce that one, and check it in Nengo to see if it's also as bad there?

**Comment by arvoelke** [comment]
Whoops. When I switched branches I accidentally wiped out my line that was fixing the vocab seed. I've edited my previous post. Think this should be consistent now with what you were seeing.

**Comment by hunse** [comment]
OK! I figured out that there's actually not a weird thing with different `dt`. It's actually just that there was a source of randomness that we were forgetting about: the splitter. I'd actually encountered this in the keyword spotting stuff and fixed it in a branch, but I didn't quite realize the importance and so I hadn't made a separate PR for it. But basically, because the splitter uses dictionaries, which do not preserve order at all, they'll randomly change the order of objects in the network, and therefore their seeds. So things were still random even when I thought they weren't! Anyway, #151 fixes this (I also pulled the fix into #132 for testing).

With this, the working memory actually works pretty consistently well. It's still not perfect, but I think it's good enough that the initial description I had for this issue no longer applies. If we want to make things better, we're going to need more specific information about where exactly working memory (and other SPA stuff) breaks down.
